File: frontend/src/models/sales/Invoice.ts
export enum InvoiceStatus {
  DRAFT = "draft",
  SENT = "sent",
  VIEWED = "viewed",
  PAID = "paid",
  PARTIALLY_PAID = "partially_paid",
  OVERDUE = "overdue",
  CANCELLED = "cancelled",
  VOID = "void",
}

export enum PaymentMethod {
  CREDIT_CARD = "credit_card",
  BANK_TRANSFER = "bank_transfer",
  CASH = "cash",
  CHECK = "check",
  PAYPAL = "paypal",
  STRIPE = "stripe",
  CREDIT = "credit",
  OTHER = "other",
}

export interface InvoiceItem {
  id: string;
  description: string;
  quantity: number;
  salePrice: number;
  discount: number;
  taxRate: number;
  taxAmount: number;
  total: number;
  unit?: string;
  productId?: string;
  productSku?: string;
  projectId?: string;
  taskId?: string;
}

export interface InvoiceItemCreate {
  description: string;
  quantity: number;
  salePrice: number;
  discount: number;
  taxRate: number;
  unit?: string;
  productId?: string;
  projectId?: string;
  taskId?: string;
}

export interface Invoice {
  id: string;
  invoiceNumber: string;
  customerId: string;
  customerName: string;
  customerEmail: string;
  customerPhone?: string;
  billingAddress: string;
  shippingAddress?: string;
  issueDate: string;
  dueDate: string;
  orderNumber?: string;
  orderTime?: string;
  subtotal: number;
  taxAmount: number;
  vatRate?: number;
  labourCost?: number;
  total: number;
  terms?: string;
  status: InvoiceStatus;
  items: InvoiceItem[];
  opportunityId?: string;
  quoteId?: string;
  projectId?: string;
  sentAt?: string;
  viewedAt?: string;
  paidAt?: string;
  overdueAt?: string;
  createdAt: string;
  updatedAt: string;
  payments: any[];
  totalPaid: number;
  balance: number;
  daysOverdue: number;
  vehicleReg?: string;
  jobCardId?: string;
}

export interface InvoiceCreate {
  customerId: string;
  customerName: string;
  customerEmail: string;
  customerPhone?: string;
  shippingAddress?: string;
  issueDate: string;
  dueDate: string;
  orderNumber?: string;
  orderTime?: string;
  labourCost?: number;
  vatRate?: number;
  terms?: string;
  items: InvoiceItemCreate[];
  opportunityId?: string;
  quoteId?: string;
  projectId?: string;
  vehicleReg?: string;
  jobCardId?: string;
}

export interface InvoiceUpdate {
  customerName?: string;
  customerEmail?: string;
  customerPhone?: string;
  shippingAddress?: string;
  issueDate?: string;
  dueDate?: string;
  orderNumber?: string;
  orderTime?: string;
  labourCost?: number;
  vatRate?: number;
  terms?: string;
  status?: InvoiceStatus;
  items?: InvoiceItemCreate[];
  vehicleReg?: string;
  jobCardId?: string;
}

export interface InvoiceFilters {
  status?: string;
  customerId?: string;
  dateFrom?: string;
  dateTo?: string;
  amountFrom?: number;
  amountTo?: number;
  search?: string;
  orderPrefix?: string;
}

File: backend/src/models/invoices/invoice.py
import uuid
from datetime import datetime
from sqlalchemy import Column, String, DateTime, Float, Text, ForeignKey, JSON, Index
from sqlalchemy.orm import relationship
from sqlalchemy.dialects.postgresql import UUID
from ...config.database_config import Base


class Invoice(Base):
    __tablename__ = "invoices"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, index=True)
    invoiceNumber = Column(String, index=True)
    tenant_id = Column(UUID(as_uuid=True), ForeignKey("tenants.id"), nullable=False)
    createdBy = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    customerId = Column(String, nullable=True)
    customerName = Column(String, nullable=False)
    customerEmail = Column(String, nullable=True)
    customerPhone = Column(String, nullable=True)
    billingAddress = Column(Text, nullable=True)
    shippingAddress = Column(Text, nullable=True)
    customerCity = Column(String, nullable=True)
    customerState = Column(String, nullable=True)
    customerPostalCode = Column(String, nullable=True)
    customerCountry = Column(String, nullable=True)
    issueDate = Column(DateTime, nullable=False)
    dueDate = Column(DateTime, nullable=False)
    orderNumber = Column(String, nullable=True)
    orderTime = Column(DateTime, nullable=True)
    subtotal = Column(Float, default=0.0)
    discountAmount = Column(Float, default=0.0)
    taxAmount = Column(Float, default=0.0)
    vatRate = Column(Float, default=0.0)
    labourCost = Column(Float, default=0.0)
    total = Column(Float, default=0.0)
    terms = Column(Text, nullable=True)
    contactId = Column(UUID(as_uuid=True), ForeignKey("contacts.id"), nullable=True)
    opportunityId = Column(String, nullable=True)
    quoteId = Column(String, nullable=True)
    projectId = Column(String, nullable=True)
    status = Column(String, default="draft")
    createdAt = Column(DateTime, default=datetime.utcnow)
    updatedAt = Column(DateTime, default=datetime.utcnow)
    paidAt = Column(DateTime, nullable=True)
    sentAt = Column(DateTime, nullable=True)
    totalPaid = Column(Float, default=0.0)
    balance = Column(Float, default=0.0)
    items = Column(JSON, default=[])
    vehicleReg = Column(String, nullable=True)
    jobCardId = Column(UUID(as_uuid=True), ForeignKey("job_cards.id"), nullable=True)
    payments = relationship("Payment", back_populates="invoice")
    tenant = relationship("Tenant", back_populates="invoices")
    creator = relationship("User", back_populates="created_invoices")

    __table_args__ = (
        Index("idx_invoices_tenant_id", "tenant_id"),
        Index("idx_invoices_status", "status"),
        Index("idx_invoices_created_at", "createdAt"),
        Index("idx_invoices_due_date", "dueDate"),
        Index("idx_invoices_customer_id", "customerId"),
        Index("idx_invoices_tenant_status", "tenant_id", "status"),
        Index("idx_invoices_tenant_due_date", "tenant_id", "dueDate"),
    )
